File: us.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace CoreTechs
{
    public static class US
    {
        public static IEnumerable<State> GetStates()
        {
            yield return new State("Federated States of Micronesia", "FM", Type.FreelyAssociatedState);
            yield return new State("Marshall Islands", "MH", Type.FreelyAssociatedState);
            yield return new State("Palau", "PW", Type.FreelyAssociatedState);
            yield return new State("District of Columbia", "DC", Type.FederalDistrict);
            yield return new State("American Samoa", "AS", Type.InsularArea);
            yield return new State("Guam", "GU", Type.InsularArea);
            yield return new State("Northern Mariana Islands", "MP", Type.InsularArea);
            yield return new State("Puerto Rico", "PR", Type.InsularArea);
            yield return new State("Virgin Islands", "VI", Type.InsularArea);
            yield return new State("Armed Forces Africa", "AE", Type.Military);
            yield return new State("Armed Forces Americas", "AA", Type.Military);
            yield return new State("Armed Forces Canada", "AE", Type.Military);
            yield return new State("Armed Forces Europe", "AE", Type.Military);
            yield return new State("Armed Forces Middle East", "AE", Type.Military);
            yield return new State("Armed Forces Pacific", "AP", Type.Military);
            yield return new State("Alabama", "AL", Type.State);
            yield return new State("Alaska", "AK", Type.State);
            yield return new State("Arizona", "AZ", Type.State);
            yield return new State("Arkansas", "AR", Type.State);
            yield return new State("California", "CA", Type.State);
            yield return new State("Colorado", "CO", Type.State);
            yield return new State("Connecticut", "CT", Type.State);
            yield return new State("Delaware", "DE", Type.State);
            yield return new State("Florida", "FL", Type.State);
            yield return new State("Georgia", "GA", Type.State);
            yield return new State("Hawaii", "HI", Type.State);
            yield return new State("Idaho", "ID", Type.State);
            yield return new State("Illinois", "IL", Type.State);
            yield return new State("Indiana", "IN", Type.State);
            yield return new State("Iowa", "IA", Type.State);
            yield return new State("Kansas", "KS", Type.State);
            yield return new State("Kentucky", "KY", Type.State);
            yield return new State("Louisiana", "LA", Type.State);
            yield return new State("Maine", "ME", Type.State);
            yield return new State("Maryland", "MD", Type.State);
            yield return new State("Massachusetts", "MA", Type.State);
            yield return new State("Michigan", "MI", Type.State);
            yield return new State("Minnesota", "MN", Type.State);
            yield return new State("Mississippi", "MS", Type.State);
            yield return new State("Missouri", "MO", Type.State);
            yield return new State("Montana", "MT", Type.State);
            yield return new State("Nebraska", "NE", Type.State);
            yield return new State("Nevada", "NV", Type.State);
            yield return new State("New Hampshire", "NH", Type.State);
            yield return new State("New Jersey", "NJ", Type.State);
            yield return new State("New Mexico", "NM", Type.State);
            yield return new State("New York", "NY", Type.State);
            yield return new State("North Carolina", "NC", Type.State);
            yield return new State("North Dakota", "ND", Type.State);
            yield return new State("Ohio", "OH", Type.State);
            yield return new State("Oklahoma", "OK", Type.State);
            yield return new State("Oregon", "OR", Type.State);
            yield return new State("Pennsylvania", "PA", Type.State);
            yield return new State("Rhode Island", "RI", Type.State);
            yield return new State("South Carolina", "SC", Type.State);
            yield return new State("South Dakota", "SD", Type.State);
            yield return new State("Tennessee", "TN", Type.State);
            yield return new State("Texas", "TX", Type.State);
            yield return new State("Utah", "UT", Type.State);
            yield return new State("Vermont", "VT", Type.State);
            yield return new State("Virginia", "VA", Type.State);
            yield return new State("Washington", "WA", Type.State);
            yield return new State("West Virginia", "WV", Type.State);
            yield return new State("Wisconsin", "WI", Type.State);
            yield return new State("Wyoming", "WY", Type.State);
        }

        public static IEnumerable<State> GetStates(params Type[] types)
        {
            return GetStates().Where(x => types.Contains(x.Type));
        }

        public static State FindByName(string name)
        {
            return GetStates().FirstOrDefault(x => x.Name.Equals(name, StringComparison.OrdinalIgnoreCase));
        }

        public static State FindByAbbreviation(string abbr)
        {
            return GetStates().FirstOrDefault(x => x.Abbreviation.Equals(abbr, StringComparison.OrdinalIgnoreCase));
        }
    }

    public class State
    {
        public State(string name, string abbr, Type type)
        {
            Name = name;
            Abbreviation = abbr;
            Type = type;
        }

        public Type Type { get; set; }
        public string Name { get; set; }
        public string Abbreviation { get; set; }
    }

    public enum Type
    {
        State,
        FederalDistrict,
        InsularArea,
        FreelyAssociatedState,
        Military
    }
}

File: us.js
var us;
(function (us) {
    var Type = (function () {
        function Type(name) {
            this.name = name;
        }
        Type.S = new Type('State');
        Type.FD = new Type('Federal District');
        Type.M = new Type('Military');
        Type.FAS = new Type('Freely Associated State');
        Type.IA = new Type('Insular Area');
        return Type;
    })();
    us.Type = Type;
    var State = (function () {
        function State(name, abbr, type) {
            this.name = name;
            this.abbr = abbr;
            this.type = type;
        }
        return State;
    })();
    us.State = State;
    us.states = [
        new State('Federated States of Micronesia', 'FM', Type.FAS),
        new State('Marshall Islands', 'MH', Type.FAS),
        new State('Palau', 'PW', Type.FAS),
        new State('District of Columbia', 'DC', Type.FD),
        new State('American Samoa', 'AS', Type.IA),
        new State('Guam', 'GU', Type.IA),
        new State('Northern Mariana Islands', 'MP', Type.IA),
        new State('Puerto Rico', 'PR', Type.IA),
        new State('Virgin Islands', 'VI', Type.IA),
        new State('Armed Forces Africa', 'AE', Type.M),
        new State('Armed Forces Americas', 'AA', Type.M),
        new State('Armed Forces Canada', 'AE', Type.M),
        new State('Armed Forces Europe', 'AE', Type.M),
        new State('Armed Forces Middle East', 'AE', Type.M),
        new State('Armed Forces Pacific', 'AP', Type.M),
        new State('Alabama', 'AL', Type.S),
        new State('Alaska', 'AK', Type.S),
        new State('Arizona', 'AZ', Type.S),
        new State('Arkansas', 'AR', Type.S),
        new State('California', 'CA', Type.S),
        new State('Colorado', 'CO', Type.S),
        new State('Connecticut', 'CT', Type.S),
        new State('Delaware', 'DE', Type.S),
        new State('Florida', 'FL', Type.S),
        new State('Georgia', 'GA', Type.S),
        new State('Hawaii', 'HI', Type.S),
        new State('Idaho', 'ID', Type.S),
        new State('Illinois', 'IL', Type.S),
        new State('Indiana', 'IN', Type.S),
        new State('Iowa', 'IA', Type.S),
        new State('Kansas', 'KS', Type.S),
        new State('Kentucky', 'KY', Type.S),
        new State('Louisiana', 'LA', Type.S),
        new State('Maine', 'ME', Type.S),
        new State('Maryland', 'MD', Type.S),
        new State('Massachusetts', 'MA', Type.S),
        new State('Michigan', 'MI', Type.S),
        new State('Minnesota', 'MN', Type.S),
        new State('Mississippi', 'MS', Type.S),
        new State('Missouri', 'MO', Type.S),
        new State('Montana', 'MT', Type.S),
        new State('Nebraska', 'NE', Type.S),
        new State('Nevada', 'NV', Type.S),
        new State('New Hampshire', 'NH', Type.S),
        new State('New Jersey', 'NJ', Type.S),
        new State('New Mexico', 'NM', Type.S),
        new State('New York', 'NY', Type.S),
        new State('North Carolina', 'NC', Type.S),
        new State('North Dakota', 'ND', Type.S),
        new State('Ohio', 'OH', Type.S),
        new State('Oklahoma', 'OK', Type.S),
        new State('Oregon', 'OR', Type.S),
        new State('Pennsylvania', 'PA', Type.S),
        new State('Rhode Island', 'RI', Type.S),
        new State('South Carolina', 'SC', Type.S),
        new State('South Dakota', 'SD', Type.S),
        new State('Tennessee', 'TN', Type.S),
        new State('Texas', 'TX', Type.S),
        new State('Utah', 'UT', Type.S),
        new State('Vermont', 'VT', Type.S),
        new State('Virginia', 'VA', Type.S),
        new State('Washington', 'WA', Type.S),
        new State('West Virginia', 'WV', Type.S),
        new State('Wisconsin', 'WI', Type.S),
        new State('Wyoming', 'WY', Type.S)
    ];
    function getAbbreviation(name) {
        name = name.toLowerCase();
        for (var i = 0; i < us.states.length; i++) {
            if (us.states[i].name.toLowerCase() === name) {
                return us.states[i].abbr;
            }
        }
    }
    us.getAbbreviation = getAbbreviation;
    function getName(abbreviation) {
        abbreviation = abbreviation.toLowerCase();
        for (var i = 0; i < us.states.length; i++) {
            if (us.states[i].abbr.toLowerCase() === abbreviation) {
                return us.states[i].name;
            }
        }
    }
    us.getName = getName;
    function getNamesAndAbbreviations() {
        return getNames.apply(this, arguments).concat(getAbbreviations.apply(this, arguments));
    }
    us.getNamesAndAbbreviations = getNamesAndAbbreviations;
    function getAbbreviations() {
        var states = getStates.apply(this, arguments);
        var names = new Array();
        for (var i = 0; i < states.length; i++) {
            names.push(states[i].abbr);
        }
        return names;
    }
    us.getAbbreviations = getAbbreviations;
    function getNames() {
        var states = getStates.apply(this, arguments);
        var names = new Array();
        for (var i = 0; i < states.length; i++) {
            names.push(states[i].name);
        }
        return names;
    }
    us.getNames = getNames;
    function getStates() {
        var results = new Array();
        if (arguments.length > 0) {
            for (var ai = 0; ai < arguments.length; ai++) {
                var type = arguments[ai];
                if (type instanceof Type) {
                    for (var si = 0; si < us.states.length; si++) {
                        var s = us.states[si];
                        if (type.name === s.type.name) {
                            results.push(s);
                        }
                    }
                }
            }
            return results;
        } else {
            return us.states;
        }
    }
    us.getStates = getStates;
})(us || (us = {}));

